kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/RecordBatchUtils.java

```diff
@@ -0,0 +1,105 @@
+                        characteristics | Spliterator.SIZED | Spliterator.SUBSIZED);
+            }
+            else {
+                spliterator = Spliterators.spliteratorUnknownSize(iterator,
+                        characteristics);
+            }
+
+            Stream<Record> stream;
+            if ((spliterator.characteristics() & Spliterator.IMMUTABLE) != 0) {
+                // Per javadoc on StreamSupport#stream(), then the spliterator is immutable use the non-Supplier factory
+                stream = StreamSupport.stream(spliterator, false);
+            }
+            else {
+                stream = StreamSupport.stream(() -> spliterator, characteristics, false);
+            }
+            return stream.onClose(iterator::close);
+        }
+        catch (RuntimeException e) {
+            iterator.close();
+            throw e;
+        }
+    }
+
+    /**
+     * Convert the given {@code recordBatch} into a {@link MemoryRecords}, applying a per-record mapping operation to each record.
+     * @param recordBatch The batch of records to convert
+     * @param mapper The mapping function to apply to the records in the batch
+     * @param resultBuffer A final buffer. This is required to encourage buffer reuse when
+     * callers make multiple invocations of this method.
+     * @return A MemoryRecords containing the mapped records
+     */
+    public static MemoryRecords toMemoryRecords(
+                                                RecordBatch recordBatch,
+                                                RecordTransform mapper,
+                                                ByteBufferOutputStream resultBuffer) {
+        return recordStream(recordBatch)
+                .collect(toMemoryRecordsCollector(recordBatch, mapper, resultBuffer));
+    }
+
+    /**
+     * Factory method for a Collector that applies a mapping function as it builds a {@link MemoryRecords}.
+     */
+    private static Collector<Record, BatchAwareMemoryRecordsBuilder, MemoryRecords> toMemoryRecordsCollector(
+                                                                                                             RecordBatch recordBatch,
+                                                                                                             RecordTransform mapper,
+                                                                                                             ByteBufferOutputStream resultBuffer) {
+        return Collector.of(
+                () -> new BatchAwareMemoryRecordsBuilder(resultBuffer).addBatchLike(recordBatch),
+                (builder, record) -> builder.appendWithOffset(
+                        mapper.transformOffset(record),
+                        mapper.transformTimestamp(record),
+                        mapper.transformKey(record),
+                        mapper.transformValue(record),
+                        mapper.transformHeaders(record)),
```

Comment:
If we're going to have `RecrodTransform#resetAfterTransform` then we need to call it here. Else we need to define what (if anything) is responsible for managing state of the `ByteBuffers` returned from the source `record`.